Add Webpack and build a browser-first version

stackmystack opened the pull request:
# About
Adds webpack and builds browser-ready js file under `build/`.

The webpack config was taken verbatim from the `eslint` project.

A new entry-point is used to generate the file: `web.js`, and its sole purpose is to prepend the `no-unsanitized/` namespace in front of the `method` and `property` rules, so we could resolve them nicely from the browser.

# Background
I must confess, I am no JS developer, but I need the rules in the browser.
`eslint` already supports webpack and builds a web version.

The produced `build/no_unsatize.js` can be loaded alongside `eslint.js`:
```html
    <script src="./eslint.js"></script>
    <script src="./no_unsanitize.js"></script>
```

And now you can do:

```js
let rules = {
  env: {
    browser: true,
  },
  rules: {
    semi: 2,
    "no-unsanitized/method": "error",
    "no-unsanitized/property": "error",
  },
};

const linter = new eslint.Linter();
linter.defineRules(no_unsanitize.rules);
console.log(linter.verify("foo.innerHTML = input.value", rules));
```

The result would be:

```js
[
    {
        "ruleId": "no-unsanitized/property",
        "severity": 2,
        "message": "Unsafe assignment to innerHTML",
        "line": 1,
        "column": 1,
        "nodeType": "AssignmentExpression",
        "endLine": 1,
        "endColumn": 28
    },
    {
        "ruleId": "semi",
        "severity": 2,
        "message": "Missing semicolon.",
        "line": 1,
        "column": 28,
        "nodeType": "ExpressionStatement",
        "messageId": "missingSemi",
        "fix": {
            "range": [
                27,
                27
            ],
            "text": ";"
        }
    }
]
```

**EDIT**: forgot the missing `linter.defineRules` call.

mozfreddyb commented:
A lot of the changes and impact and warnings that I see from your pull request stem from, let's say, noisy changes: Indentation, package updates, dependencies being added etc. It's a bit hard for me to review and assess the actual impact of your change with this noise. I suggest you try fix this warnings and this will make it easier for me to discuss possible next steps. 
That being said, I want to acknowledge that making the change easier to review is a bit of work and we might still not want to take your change after all.

You said you're not an experienced JS developer and that's fine. We can happily keep that PR open for a longer while, if it takes a bit of work. Please also note that you'll be able to iterate more quickly if you run the tests locally. They are (relatively) fast: `npm run lint` to check for syntax issues and then `npm run test` to make sure our test suite passes.

stackmystack commented:
Yes they're all because of my editor not understanding the JS environment. They're all indentation issues.

Everything fixed, tests are passing, and I forced pushed this branch for you to review.

Cheers.

mozfreddyb commented:
Sorry for the delays. I am basically maintaining this project in my spare time. My actual job duties are currently shifting away from this project.

Review-wise this look good. Can you deploy this in a personal repository (branch) of yours for easier testing from my end? Thanks!

stackmystack commented:
No problem at all. But I didn't get your comment:

>  Can you deploy this in a personal repository (branch) of yours for easier testing from my end?

It's already on my personal branch [in my repo](https://github.com/stackmystack/eslint-plugin-no-unsanitized/tree/webpack-generation).

mozfreddyb commented:
Hey, sorry about the long delay. I think this is mostly good to go.
Before I merge, I also asked for review from another collaborator.

@stackmystack Can you also add `webpack build` as a separate CI step in our github actions workflow config? This will help us make sure the build does not break in upcoming changes

mozfreddyb reviewed (changes requested):
As mentioned before, please add the workflow changes to include the webpack build step

stackmystack commented:
I renamed the `webpack` step to `build` because you may want to use another tool in the future, so now it's tooling agnostic.

I also rebased, then added the requested step to the CI, and [ran it in my repo](https://github.com/stackmystack/eslint-plugin-no-unsanitized/actions/runs/4808263425). It passes.

mozfreddyb commented:
@rpl Please review

rpl commented on package.json:
```diff
@@ -9,16 +9,21 @@
         "nyc": "^15.1.0",
-        "typescript": "^3.9.7"
+        "typescript": "^3.9.7",
+        "webpack": "^5.75.0",
```
**Issue**: this webpack version does have a security advisory associated to it, I haven't checked it by the most recent version isn't affected and so we should just update the version here.

Nit: and while we are here let's also make sure webpack-cli version is the last version released too.

rpl commented on docs/rules/customization.md:
```diff
@@ -86,5 +86,5 @@ TBD
 
 #### More
-* See [our rule schema definition](SCHEMA.md). 
+* See [our rule schema definition](https://github.com/mozilla/eslint-plugin-no-unsanitized/blob/main/SCHEMA.md). 
```
Nit, This change doesn't seem to be relevant to this PR not necessary.

stackmystack replied:
I think that was added by some stupid tool. Removing.

stackmystack replied:
That's not my work, (git blames it on Alvaro Junqueira). Will remove anyway.

stackmystack replied:
On second thought, I'll keep it.

rpl commented on package.json:
```diff
@@ -41,7 +46,8 @@
         "test": "nyc mocha tests/rules/",
-        "lint": "eslint ."
+        "lint": "eslint .",
+        "build": "webpack --output-path=build"
```
May be good to clear the bulid dir as part of the build script, to make sure that if the file exists than it is for sure a fresh built one

rpl commented on webpack.config.js:
```diff
@@ -0,0 +1,53 @@
+                        ["@babel/preset-env", {
+                            debug: true, // ← to print actual browser versions
+
+                            /*
+               * We want to remove `transform-unicode-regex` convert because of https://github.com/eslint/eslint/pull/12662.
+               *
+               * With `>0.5%`, `@babel/preset-env@7.7.6` prints below:
+               *
+               *     transform-unicode-regex { "chrome":"49", "ie":"11", "safari":"5.1" }
+               *
+               * So this excludes those versions:
+               *
+               * - IE 11
+               * - Chrome 49 (2016; the last version on Windows XP)
+               * - Safari 5.1 (2011-2013; the last version on Windows)
+               */
```
I'm sure that for the people developing eslint this big inline comment in their webpack config file has a meaning and is likely pointing something they would like to happen and change this preset-env config accordingly, but for us in this repo it may not be providing that much value in its current form, instead we may want to just point to our future selves that this preset-env config is meant to be possibly kept in sync with the one in the eslint repo.

e.g. something like:

```
// NOTE: keep the preset-env targets in sync with the one
// used in https://github.com/eslint/eslint 
targets: ">0.5%, not chrome 49, not ie 11, not safari 5.1",
```

This is shorter and it does tell us what we should be possibly doing about it over the time.

rpl commented on webpack.config.js:
```diff
@@ -0,0 +1,53 @@
+            },
+        ],
+    },
+    plugins: [
+        new NodePolyfillPlugin(),
+    ],
```
Node-polyfill-webpack-plugin doesn’t seem to be polyfilling anything in the file webpack is building from the no-unsanitized sources, at least not at the moment, which is not super surprising given that we don’t use any special from nodejs that isn’t available in the js environment that webbrowsers provides.

I'd suggest to remove it all together (both the addition to the npm dependencies and from this webpack config), and spare ourselves a new dependency if not really used at the moment (and eventually reconsider if we would ever need it).

rpl commented on webpack.config.js:
```diff
@@ -0,0 +1,53 @@
+        new NodePolyfillPlugin(),
+    ],
+    resolve: {
+        mainFields: ["browser", "main", "module"],
```
If I'm not mistaken [resolve.mainFields](https://webpack.js.org/configuration/resolve/#resolvemainfields) should be instructing webcompat about which package.json property to prefer when importing dependency into a webcompat bundle, and so even if we don't expose any of those properties (and I'm not sure it does make any difference for the bundle generated from this repo at the moment) it may be worth leaving a comment to explain what it does and don’t mistake it as related to the package.json in this repo (where we only export a main property and none of the others).

It may not be doing much at the moment but I think it may be reasonable to keep it given that it doesn’t require a new dependency but just webpack configuration

rpl commented on webpack.config.js:
```diff
@@ -0,0 +1,53 @@
+const NodePolyfillPlugin = require("node-polyfill-webpack-plugin");
+
+/** @type {import("webpack").Configuration} */
+module.exports = {
+    mode: "none",
```
Question: Do we want to include devtool: "source-map" (or another variant suggested for production builds, see [webcompat docs here](https://webpack.js.org/configuration/devtool/#production)) to ensure webcompat is going to build a source map file?

mozfreddyb replied:
If it's easy to include, then let's. Otherwise, OK not to bother.

rpl commented on webpack.config.js:
```diff
@@ -0,0 +1,53 @@
+module.exports = {
+    mode: "none",
+    entry: {
+        no_unsanitize: ["./web.js"],
```
Question:

Its unfortunate we have to defined a separate entry point file web.js, but based on the eslint API docs it seems that while the ESLint class expects a plugin config property to associate plugin name with its exported object, but that is preprocessed internally before passing it to the Linter class and so Linter defineRules method seems the best workaround available at the moment (at least based on a brief digging into eslint internals it appears to be).

Still, I’m wondering (and so I wanted to ask this explicitly) if as an alternative it may be reasonable to export a new `registerRulesToLinter` method from the index.js file, which given a `Linter` instance would be taking care of registering the rules with the right namespace prefixed to each plugin rule id, e.g. by adding to index.js something like:

```
registerRulesToLinter(linter) { 
    for (const ruleId of Object.keys(this.rules)) {     
        linter.defineRule(`no-unsanitized/${ruleId}`, this.rules[ruleId]);
         }
},
```

And then using it like:

```
const linter = window.linter = new eslint.Linter();
no_unsanitized.registerRulesToLinter(linter);
```

wdyt?

stackmystack replied:
I am not sure about it, but I would absolutely hate to keep track of eslint internals and depend on them. I'd rather stick with their public API.

mozfreddyb replied:
100% agree. I'm very OK with a level of indirection if this makes maintenance less painful.

rpl commented on .github/workflows/ci.yml:
```diff
@@ -28,6 +28,9 @@ jobs:
       - name: Install
         run: yarn install
 
+      - name: Build
```
naming it "Build" may be slightly confusing for our future selves, as it may suggest that building is a necessary step in our CI, while this is mainly smoke testing that webpack building step works, but we are not actually shipping the built file in the npm package we are releasing and so I'd name it "Build webpack bundle for the web" or "Smoke test webpack bundled module for the web" to make that more immediately clear.

rpl reviewed (changes requested):
@stackmystack thanks a lot for this effort!

This PR looks largely good to me, but I have collected a few comments and a few small requests for changes (some are more of nits) and some questions.

In addition to that I was wondering:
- if we should consider adding the bundled file in our npm package or not (eslint doesn't seem to be included its webpack bundle neither to be honest and so even if we include our bundled file it may still not be that helpful for making the entire setup to be easier in practice)

- things that we release without test coverage have the annoying tendency of breaking without letting us know :-), and so it would have been nice if we could have covered this with even just a single small smoke test loading and exercising the bundled plugin to confirm it does work as expected, but the fact that eslint doesn't include their bundle in their npm package would make that test a bit more annoying to write. 

e.g. we would need to do a shallow clone of eslint repo, install its dependency and build and bundle, then assemble that with our own bundled file and run the smoke test, e.g. inside a jsdom-based mocked browser environment)

What is described above is not impossible, it is pretty doable, but require much more work than it would have required if we could have copied eslint built bundle from the eslint package instead of having to build it as part of such a test case.

And so I wouldn't block on that test case. @mozfreddyb wdyt? does that sound reasonable to you too? (don't block on the additional test case with the rationale described above)

Instead I'd say we could still add a small section in the README.md file to explain how to use the bundled file, given that requires some knowledge to work correctly (and differs enough from how you would setup the plugin when using eslint from nodejs and execute it from the command line). @mozfreddyb wdyt? you would be willing to include a brief section about that in the repo README.md file?

mozfreddyb commented:
>     if we should consider adding the bundled file in our npm package or not (eslint doesn't seem to be included its webpack bundle neither to be honest and so even if we include our bundled file it may still not be that helpful for making the entire setup to be easier in practice)
> 

No, not yet.
My main use case was really just to have a demo that people can use to quickly test the rule and file bugs. I think this could live somewhere e.g., at https://mozilla.github.ui/eslint-plugin-no-unsanitized/demo - I'd rather not make this part of what we ship.

>     things that we release without test coverage have the annoying tendency of breaking without letting us know :-), and so it would have been nice if we could have covered this with even just a single small smoke test loading and exercising the bundled plugin to confirm it does work as expected, but the fact that eslint doesn't include their bundle in their npm package would make that test a bit more annoying to write.
> 
> e.g. we would need to do a shallow clone of eslint repo, install its dependency and build and bundle, then assemble that with our own bundled file and run the smoke test, e.g. inside a jsdom-based mocked browser environment)
> 
> What is described above is not impossible, it is pretty doable, but require much more work than it would have required if we could have copied eslint built bundle from the eslint package instead of having to build it as part of such a test case.
> 
> And so I wouldn't block on that test case. @mozfreddyb wdyt? does that sound reasonable to you too? (don't block on the additional test case with the rationale described above)
> 

One of the reasons I'd want us to start off with a "demo" use case rather than making this a supported thing we ask people to use.


> Instead I'd say we could still add a small section in the README.md file to explain how to use the bundled file, given that requires some knowledge to work correctly (and differs enough from how you would setup the plugin when using eslint from nodejs and execute it from the command line). @mozfreddyb wdyt? you would be willing to include a brief section about that in the repo README.md file?

I'm not sure that's even necessary in case we just use it "internally" as suggested above.

rpl commented:
> > ```
> > if we should consider adding the bundled file in our npm package or not (eslint doesn't seem to be included its webpack bundle neither to be honest and so even if we include our bundled file it may still not be that helpful for making the entire setup to be easier in practice)
> > ```
> 
> No, not yet. My main use case was really just to have a demo that people can use to quickly test the rule and file bugs. I think this could live somewhere e.g., at https://mozilla.github.ui/eslint-plugin-no-unsanitized/demo - I'd rather not make this part of what we ship.
> 
> > ```
> > things that we release without test coverage have the annoying tendency of breaking without letting us know :-), and so it would have been nice if we could have covered this with even just a single small smoke test loading and exercising the bundled plugin to confirm it does work as expected, but the fact that eslint doesn't include their bundle in their npm package would make that test a bit more annoying to write.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > e.g. we would need to do a shallow clone of eslint repo, install its dependency and build and bundle, then assemble that with our own bundled file and run the smoke test, e.g. inside a jsdom-based mocked browser environment)
> > What is described above is not impossible, it is pretty doable, but require much more work than it would have required if we could have copied eslint built bundle from the eslint package instead of having to build it as part of such a test case.
> > And so I wouldn't block on that test case. @mozfreddyb wdyt? does that sound reasonable to you too? (don't block on the additional test case with the rationale described above)
> 
> One of the reasons I'd want us to start off with a "demo" use case rather than making this a supported thing we ask people to use.
> 
> > Instead I'd say we could still add a small section in the README.md file to explain how to use the bundled file, given that requires some knowledge to work correctly (and differs enough from how you would setup the plugin when using eslint from nodejs and execute it from the command line). @mozfreddyb wdyt? you would be willing to include a brief section about that in the repo README.md file?
> 
> I'm not sure that's even necessary in case we just use it "internally" as suggested above.

@mozfreddyb sounds good to me, as I mentioned in my comments I was already considering all of those as non-blocking issues from my perspective, let's just aim to cover the few small tweaks proposed and then merge this.

mozfreddyb commented:
@stackmystack Sorry it took us so long. Are you still interested in maintaining this Pull Request and bringing it to completion?

stackmystack commented:
> @stackmystack Sorry it took us so long. Are you still interested in maintaining this Pull Request and bringing it to completion?

Hello @mozfreddyb, I am on vacation, and will not be available before 5 weeks. So If you can wait until then, I'd happily spare some time to maintain it. But I guess you guys agreed on little-to-no-modifications AFAICT, so If you feel you want to merge early, please be my guest, the branch is yours :)

mozfreddyb commented:
@stackmystack Please take a look at some of the suggested changes. We can easily wait ~5 more weeks :)

stackmystack commented:
@mozfreddyb @rpl 

I integrated as much of your comments and suggestions as possible.

Ready for a second iteration of review, since I added comments/skipped some suggestions.

I will also skip the smoke test part since it's a lot of effor given that my day job requires more attention on different topics, so it wouldn't be nice of me to promise I'd do it when I can't be sure about it.
